supplier_invoice_number_unique/model/account_invoice.py

```diff
@@ -40,8 +40,24 @@ def _compute_supplier_invoice_number_strip(self):
             invoice.supplier_invoice_number_strip = re.sub(
                 r'\W+|\_', '', invoice.supplier_invoice_number.lower())
 
+    @api.multi
+    def invoice_validate(self):
+        constraint = self.env.context.get('constraint', 'DEFERRED')
+        query = """
+        SET CONSTRAINTS
+            account_invoice_unique_supplier_invoice_number_strip DEFERRED
+        """
+        if constraint != 'DEFERRED':
+            query = """
+            SET CONSTRAINTS
+                account_invoice_unique_supplier_invoice_number_strip IMMEDIATE
+            """
+        self.env.cr.execute(query)
```

Comment:
i don't understand why you are changing the constraint type at the moment  to validate the invoice

Reply:
because the constraint are not raised with the odoo warning, is raised like a traceback

![image](https://user-images.githubusercontent.com/10233975/31236739-9835c07e-a9ba-11e7-9156-e8917a2111c6.png)

So, in order to get the raised odoo warning I change the mode of sql constraint

Reply:
@hugho-ad 
commonly the sql constraint return warning
do you know why in this case return traceback, maybe because is trigger by function fields?

Reply:
I think that the most probable cause is  because it is trigger by the work flow, instead of a direct write over the record, (save button on the form view or simillar)

Reply:
@hugho-ad let's wait a little bit for the review of @JulioSerna he need to make a test but is really busy at this moment, he will let us know asap.

Reply:
@hugho-ad 

I make some test and I think is by the function fields in the test I changed the field by char with onchange

```
         @api.onchange('supplier_invoice_number', 'state', 'type', 'fiscal_position')
         def onchange_number(self):
             self.supplier_invoice_number_strip = re.sub(
                 r'\W+|\_', '', self.supplier_invoice_number.lower())
```

and the warning is returned

Could you check if this is happening in version11.0 to report the issue?

about your pr what about make the constraint IMMEDIATE in order to do not change in invoice_validation method, but I do not the impact of this change as in the issue comment

>Doing this appears to fully work round the issue - but in more complex scenarios I have a feeling this would blow up, as it's a pretty sketchy workaround

what do you think @moylop260 ?

Reply:
@JulioSerna 
> I make some test and I think is by the function fields in the test I changed the field by char with onchange and the warning is returned

Instead a computeted field, do you change the field in order to made it  a normal field with a onchange function?

Also The field must  be in the view in order to make works the sql_constraint, I also made the test, this can work just keeping the field invisible, what do you think @dsabrinarg, this works only when you are edditing the record and saving it or triggering the workflow while the form view is in edit mode

The difference is that witht the onchange function, the field `supplier_invoice_number_strip` gonna be calculated in each editing record from the form view

> about your pr what about make the constraint IMMEDIATE in order to do not change in invoice_validation method, 

I already try it, making the constraint IMMEDIATE, the warinig error is NOT displayed and instead of it, appears the traceback error

> but I do not the impact of this change as in the issue comment

in the  issue,  PeterAlabaster  comments that because he is making a `cr.commit` in the writte method.

> Could you check if this is happening in version11.0 to report the issue?

Sure, I test it already, and the warning error is triggered properly does not matter if it trigger from a computed field, but I was not able to tets it from a workflow

I test it with the following code:
```python
test2 = fields.Char(compute="_compute_test_field", store=True)                  
test = fields.Char()                                                            
                                                                                      
      @api.depends('test')                                                                                                                                                                                                                     
      def _compute_test_field(self):                                                  
          for record in self:                                                         
              if record.test == 'test':                                               
                  record.test2 = record.test                                          
                                                                                      
      _sql_constraints = [                                                            
          ('test_uniq', 'unique(partner_id, test2)', 'TEST!'),                        
      ]                                                
```

Reply:
@dsabrinarg @JulioSerna @hugho-ad 

I don't like this solution.

If the problem is [this](https://github.com/Vauxoo/addons-vauxoo/pull/1233/files#r142979576). I rather use something like this(While we reporting the error to odoo)

```python
def invoice_validate(self):
    try:
        res = super(AccountInvoice, self).invoice_validate()
    except ValueError, msg:
        raise BEAUTIFUL_HUGO_MSG
    return res
```

What do you think?

Best Regards.